File: Trans_PY/Train_TransE.py
import random
from math import sqrt
from numpy import mat
from numpy.linalg import norm
import pickle
import numpy
import datetime

datapath = "../FB15k/"

entity2id = {}
id2entity = {}
relation2id = {}
id2relation = {}
fb_h = []
fb_t = []
fb_r = []
entity_num = 0
relation_num = 0
relation_vec = []
entity_vec = []
ok = {}

dim = 100
rate = 0.01
margin = 1.0
method = "bern"

L1_FLAG = True


def normalize(miu, sigma, mimi, maxi):
	n = random.gauss(miu, sigma)
	while n < mimi or n > maxi:
		n = (maxi - mimi) * random.random() + random.gauss(miu, sigma)
	return n

def parseline(line):
	vec = []
	for i in line.split("\t"):
		if i != "\n":
			vec += [float(i)]
	return vec

def train_kb(e1_a, e2_a, rel_a, e1_b, e2_b, rel_b):
	global res, count
	sum1 = calc_sum(e1_a, e2_a, rel_a)
	sum2 = calc_sum(e1_b, e2_b, rel_b)
	if sum1+margin>sum2:
		count += 1
		res = res + margin + sum1 - sum2
	#	s = datetime.datetime.now()
		gradient(e1_a, e2_a, rel_a, e1_b, e2_b, rel_b)
	#	e = datetime.datetime.now()
	#	print "gradient: ", (e - s).microseconds
def calc_sum(e1, e2, rel):
	sum1 = mat(entity_vec[e2]) - mat(entity_vec[e1]) - mat(relation_vec[rel])
	sum1 = sum1.getA()[0]
	if L1_FLAG:
	#	for i in xrange(dim):
	#		sum1 = sum1 + abs(entity_vec[e2][i]-entity_vec[e1][i]-relation_vec[rel][i])
		sum1 = sum(abs(sum1))
	else:
	#	for i in xrange(dim):
	#		sum1 = sum1 + (entity_vec[e2][i]-entity_vec[e1][i]-relation_vec[rel][i])**2
		sum1 = sum(sum1**2)
	return sum1

def gradient(e1_a, e2_a, rel_a, e1_b, e2_b, rel_b):
	global relation_tmp, entity_tmp
#	for i in xrange(dim):
#		x = 2*(entity_vec[e2_a][i]-entity_vec[e1_a][i]-relation_vec[rel_a][i])
#		if L1_FLAG:
#			if x > 0:
#				x = 1
#			else:
#				x = -1
#		relation_tmp[rel_a][i] = relation_tmp[rel_a][i] + rate*x
#		entity_tmp[e1_a][i] = entity_tmp[e1_a][i] + rate*x
#		entity_tmp[e2_a][i] = entity_tmp[e2_a][i] - rate*x
#		x = 2*(entity_vec[e2_b][i]-entity_vec[e1_b][i]-relation_vec[rel_b][i])
#		if L1_FLAG:
#			if x > 0:
#				x = 1
#			else:
#				x = -1
#		relation_tmp[rel_b][i] = relation_tmp[rel_b][i] - rate*x
#		entity_tmp[e1_b][i] = entity_tmp[e1_b][i] - rate*x
#		entity_tmp[e2_b][i] = entity_tmp[e2_b][i] + rate*x
	x = 2*(mat(entity_vec[e2_a]) - mat(entity_vec[e1_a]) - mat(relation_vec[rel_a]))
	x = x.getA()[0]
	if L1_FLAG:
		x[x > 0] = 1
		x[x <= 0] = -1
	relation_tmp[rel_a] = relation_tmp[rel_a] + rate*x
	entity_tmp[e1_a] = entity_tmp[e1_a] + rate*x
	entity_tmp[e2_a] = entity_tmp[e2_a] - rate*x
	x = 2*(mat(entity_vec[e2_b]) - mat(entity_vec[e1_b]) - mat(relation_vec[rel_b]))
	x = x.getA()[0]
	if L1_FLAG:
		x[x > 0] = 1
		x[x <= 0] = -1
	relation_tmp[rel_b] = relation_tmp[rel_b] - rate*x
	entity_tmp[e1_b] = entity_tmp[e1_b] - rate*x
	entity_tmp[e2_b] = entity_tmp[e2_b] + rate*x
# perpare function
f = open(datapath+"entity2id.txt","r")
data = f.readlines()
f.close()
for i in data:
	str1, ind = i.split("\t")
	str1 = str1.strip()
	entity2id[str1] = int(ind)
	id2entity[int(ind)] = str1
	entity_num = entity_num + 1

f = open(datapath+"relation2id.txt","r")
data = f.readlines()
f.close()
for i in data:
	str1, ind = i.split("\t")
	str1 = str1.strip()
	relation2id[str1] = int(ind)
	id2relation[int(ind)] = str1
	relation_num = relation_num + 1
#left_entity = [[0 for i in xrange(entity_num)] for i in xrange(relation_num)]
#right_entity = [[0 for i in xrange(entity_num)] for i in xrange(relation_num)]
left_entity = {}
right_entity = {}
left_num = [0 for i in xrange(relation_num)]
right_num = [0 for i in xrange(relation_num)]

f = open(datapath+"train.txt", "r")
data = f.readlines()
f.close()
for i in data:
	left, right, rel = i.split("\t")
	left = left.strip()
	right = right.strip()
	rel = rel.strip()
	if(entity2id.has_key(left) == False):
		print "miss entity:", left
	if(entity2id.has_key(right) == False):
		print "miss entity:", right
	if(relation2id.has_key(rel) == False):
		relation2id[rel] = relation_num
		relation_num = relation_num + 1
		print "miss relationship: ", rel
	if left_entity.has_key(relation2id[rel]) and left_entity[relation2id[rel]].has_key(entity2id[left]):
		left_entity[relation2id[rel]][entity2id[left]] = left_entity[relation2id[rel]][entity2id[left]] + 1
	elif left_entity.has_key(relation2id[rel]):
		left_entity[relation2id[rel]][entity2id[left]] = 1
	else:
		left_entity[relation2id[rel]] = {entity2id[left]:1}
	if right_entity.has_key(relation2id[rel]) and right_entity[relation2id[rel]].has_key(entity2id[right]):
		right_entity[relation2id[rel]][entity2id[right]] = right_entity[relation2id[rel]][entity2id[right]] + 1
	elif right_entity.has_key(relation2id[rel]):
		right_entity[relation2id[rel]][entity2id[right]] = 1
	else:
		right_entity[relation2id[rel]] = {entity2id[right]:1}
	fb_h += [entity2id[left]]
	fb_t += [entity2id[right]]
	fb_r += [relation2id[rel]]
	#ok = {entity2id[left]:{relation2id[rel]:{entity2id[right]:1}}}
	ok[(str(entity2id[left])+"@"+str(relation2id[rel])+"@"+str(entity2id[right]))] = 1
for key, value in left_entity.items():
	sum1 = 0
	sum2 = 0
	for v in value.values():
		sum1 = sum1 + 1
		sum2 = sum2 + v
	left_num[key] = 1.0*sum2/sum1

for key, value in right_entity.items():
	sum1 = 0
	sum2 = 0
	for v in value.values():
		sum1 = sum1 + 1
		sum2 = sum2 + v
	right_num[key] = 1.0*sum2/sum1

# run function
#entity_vec = mat([[random.uniform(-6.0/sqrt(dim), 6.0/sqrt(dim)) for i in xrange(dim)] for i in xrange(entity_num)])
#relation_vec = mat([[random.uniform(-6.0/sqrt(dim), 6.0/sqrt(dim)) for i in xrange(dim)] for i in xrange(relation_num)])
#entity_vec = [[random.uniform(-6.0/sqrt(dim), 6.0/sqrt(dim)) for i in xrange(dim)] for i in xrange(entity_num)]
#relation_vec = [[random.uniform(-6.0/sqrt(dim), 6.0/sqrt(dim)) for i in xrange(dim)] for i in xrange(relation_num)]
#entity_vec = [[random.gauss(0, 1.0/dim) for i in xrange(dim)] for i in xrange(entity_num)]
#relation_vec = [[random.gauss(0, 1.0/dim) for i in xrange(dim)] for i in xrange(relation_num)]
entity_vec = [[normalize(0,1.0/dim, -6.0/sqrt(dim), 6.0/sqrt(dim)) for i in xrange(dim)] for i in xrange(entity_num)]
relation_vec = [[normalize(0,1.0/dim, -6.0/sqrt(dim), 6.0/sqrt(dim)) for i in xrange(dim)] for i in xrange(relation_num)]

for i in xrange(len(entity_vec)):
	entity_vec[i] /= norm(entity_vec[i])
	#entity_vec[i] = entity_vec[i] / len(entity_vec[i])

#bfgs function
res = 0
nbatches=100
nepoch=1000
batchsize=len(fb_h)/nbatches
cnt = 0
rate_flag = []

print "relation_num: ", relation_num, "entity_num: ", entity_num
for epoch in xrange(nepoch):
	res = 0
	count = 0
	start = datetime.datetime.now()
	for batch in xrange(nbatches):
		relation_tmp = relation_vec
		entity_tmp = entity_vec
#		batch_start = datetime.datetime.now()
		for k in xrange(batchsize):
			i = (random.randint(0, len(fb_h)-1) * random.randint(0, len(fb_h)-1))%len(fb_h)
			j = (random.randint(0, entity_num-1) * random.randint(0, entity_num-1))%entity_num
			pr = right_num[fb_r[i]]/(right_num[fb_r[i]]+left_num[fb_r[i]])
			if(method=="unif"):
				pr = 0.5
			if(random.uniform(0, 1)<pr):
				while ok.has_key((str(fb_h[i])+"@"+str(fb_r[i])+"@"+str(j))):
					j = (random.randint(0, entity_num-1) * random.randint(0, entity_num-1))%entity_num
#				s = datetime.datetime.now()
				train_kb(fb_h[i], fb_t[i], fb_r[i], fb_h[i], j, fb_r[i])
#				e = datetime.datetime.now()
#				print "train_kb:", (e - s).microseconds
			else:
				while ok.has_key((str(j)+"@"+str(fb_r[i])+"@"+str(fb_t[i]))):
					j = (random.randint(0, entity_num-1) * random.randint(0, entity_num-1))%entity_num
				train_kb(fb_h[i], fb_t[i], fb_r[i], j, fb_t[i], fb_r[i])
	                relation_tmp[fb_r[i]] /=  norm(relation_tmp[fb_r[i]])
			entity_tmp[fb_h[i]] /= norm(entity_tmp[fb_h[i]])
			entity_tmp[fb_t[i]] /= norm(entity_tmp[fb_t[i]])
			entity_tmp[j] /= norm(entity_tmp[j])
		entity_vec = entity_tmp
		relation_vec = relation_tmp
#		batch_end = datetime.datetime.now()
#		print "batch ", batch, (batch_end - batch_start).microseconds
	end = datetime.datetime.now()
	print "epoch:", epoch, res, "time: ", (end - start).seconds, " rate: ", rate, "gradient:", count
	if rate_flag:
		tmp = numpy.mean(rate_flag)
		if res < tmp and (tmp - res)/tmp > 0.01:
			rate_flag = []
		else:
			if cnt > 5:
				rate_flag = []
				rate = rate / 10.0
				if rate < 0.0001:
					rate = 0.001
				cnt = 0
				print "change rate: ", rate
			else:
				rate_flag += [res]
				cnt = cnt + 1
        else:
		rate_flag += [res]
	f1 = open("relation2vec."+method, "w")
	f2 = open("entity2vec."+method, "w")
	pickle.dump(relation_vec, f1)
	pickle.dump(entity_vec, f2)
	f1.close()
	f2.close()

File: Trans_PY/TransN/Train_TransE.cpp
#include<iostream>
#include<cstring>
#include<cstdio>
#include<map>
#include<vector>
#include<string>
#include<ctime>
#include<cmath>
#include<cstdlib>
using namespace std;


#define pi 3.1415926535897932384626433832795

bool L1_flag=1;

//normal distribution
double rand(double min, double max)
{
    return min+(max-min)*rand()/(RAND_MAX+1.0);
}
double normal(double x, double miu,double sigma)
{
    return 1.0/sqrt(2*pi)/sigma*exp(-1*(x-miu)*(x-miu)/(2*sigma*sigma));
}
double randn(double miu,double sigma, double min ,double max)
{
    double x,y,dScope;
    do{
        x=rand(min,max);
        y=normal(x,miu,sigma);
        dScope=rand(0.0,normal(miu,miu,sigma));
    }while(dScope>y);
    return x;
}

double sqr(double x)
{
    return x*x;
}

double vec_len(vector<double> &a)
{
	double res=0;
    for (int i=0; i<a.size(); i++)
		res+=a[i]*a[i];
	res = sqrt(res);
	return res;
}

string version;
string data_path = "../../FB15k/";
char buf[100000],buf1[100000];
int relation_num,entity_num;
map<string,int> relation2id,entity2id;
map<int,string> id2entity,id2relation;


map<int,map<int,int> > left_entity,right_entity;
map<int,double> left_num,right_num;
class Train{

public:
	map<pair<int, int>, vector<int> >entity_neighbors;
	map<pair<int,int>, map<int,int> > ok;
    void add(int x,int y,int z)
    {
        fb_h.push_back(x);
        fb_r.push_back(z);
        fb_l.push_back(y);
        ok[make_pair(x,z)][y]=1;
	entity_neighbors[make_pair(x,z)].push_back(y);
//	if entity_neighbors.count(make_pair(x,z)) > 0
//		entity_pair[make_pair(x,z)].push_back(y);
//	else
//		entity_pair[make_pair(x,z)].push_back(y);
    }
    void run(int n_in,double rate_in,double margin_in,int method_in, double con_alpha_in)
    {
        n = n_in;
        rate = rate_in;
        margin = margin_in;
        method = method_in;
	con_alpha = con_alpha_in;
        relation_vec.resize(relation_num);
		for (int i=0; i<relation_vec.size(); i++)
			relation_vec[i].resize(n);
        entity_vec.resize(entity_num);
		for (int i=0; i<entity_vec.size(); i++)
			entity_vec[i].resize(n);
        relation_tmp.resize(relation_num);
		for (int i=0; i<relation_tmp.size(); i++)
			relation_tmp[i].resize(n);
        entity_tmp.resize(entity_num);
		for (int i=0; i<entity_tmp.size(); i++)
			entity_tmp[i].resize(n);
	for (int i=0; i<relation_num; i++)
        {
            for (int ii=0; ii<n; ii++)
                relation_vec[i][ii] = randn(0,1.0/n,-6/sqrt(n),6/sqrt(n));
        }
        for (int i=0; i<entity_num; i++)
        {
            for (int ii=0; ii<n; ii++)
                entity_vec[i][ii] = randn(0,1.0/n,-6/sqrt(n),6/sqrt(n));
            norm(entity_vec[i]);
        }
	
        bfgs();
    }

private:
    int n,method;
    double res;//loss function value
    double count,count1;//loss function gradient
    double rate,margin, con_alpha;
    double belta;
    vector<int> fb_h,fb_l,fb_r;
    vector<vector<int> > feature;
    vector<vector<double> > relation_vec,entity_vec;
    vector<vector<double> > relation_tmp,entity_tmp;
    double norm(vector<double> &a)
    {
        double x = vec_len(a);
        if (x>1)
        for (int ii=0; ii<a.size(); ii++)
                a[ii]/=x;
        return 0;
    }
    int rand_max(int x)
    {
        int res = (rand()*rand())%x;
        while (res<0)
            res+=x;
        return res;
    }

    void bfgs()
    {
        res=0;
        int nbatches=100;
        int nepoch = 1000;
        int batchsize = fb_h.size()/nbatches;
            for (int epoch=0; epoch<nepoch; epoch++)
            {

            	res=0;
		time_t epoch_s, epoch_e;
		epoch_s = time(NULL);
             	for (int batch = 0; batch<nbatches; batch++)
             	{
			time_t start = time(NULL), stop;
			int tmp_res = res;
             		relation_tmp=relation_vec;
            		entity_tmp = entity_vec;
             		for (int k=0; k<batchsize; k++)
             		{
						int i=rand_max(fb_h.size());
						int j=rand_max(entity_num);
						double pr = 1000*right_num[fb_r[i]]/(right_num[fb_r[i]]+left_num[fb_r[i]]);
						if (method ==0)
			                            pr = 500;
						if (rand()%1000<pr)
						{
							while (ok[make_pair(fb_h[i],fb_r[i])].count(j)>0)
								j=rand_max(entity_num);
							train_kb(fb_h[i],fb_l[i],fb_r[i],fb_h[i],j,fb_r[i]);
						}
						else
						{
							while (ok[make_pair(j,fb_r[i])].count(fb_l[i])>0)
								j=rand_max(entity_num);
							train_kb(fb_h[i],fb_l[i],fb_r[i],j,fb_l[i],fb_r[i]);
						}
                		norm(relation_tmp[fb_r[i]]);
                		norm(entity_tmp[fb_h[i]]);
                		norm(entity_tmp[fb_l[i]]);
                		norm(entity_tmp[j]);
             		
			}
			stop = time(NULL);
			cout << "batch: " << batch << " res: " << res - tmp_res << " time" << (double)(stop-start) << endl;
		        relation_vec = relation_tmp;
		        entity_vec = entity_tmp;
             	}
		epoch_e = time(NULL);
                cout<<"epoch:"<<epoch<<' '<<res<<(double)(epoch_e-epoch_s)<<endl;
                FILE* f2 = fopen(("relation2vec."+version).c_str(),"w");
                FILE* f3 = fopen(("entity2vec."+version).c_str(),"w");
		if(f2 ==  NULL || f3 == NULL)
		   cout << "file open failed" << endl;
                for (int i=0; i<relation_num; i++)
                {
                    for (int ii=0; ii<n; ii++)
                        fprintf(f2,"%.6lf\t",relation_vec[i][ii]);
                    fprintf(f2,"\n");
                }
                for (int i=0; i<entity_num; i++)
                {
                    for (int ii=0; ii<n; ii++)
                        fprintf(f3,"%.6lf\t",entity_vec[i][ii]);
                    fprintf(f3,"\n");
                }
                fclose(f2);
                fclose(f3);
            }
    }
    double res1;
    vector<double> get_neighbors_attention_vec(int e, int rel)
    {
	vector<double> attention_vec(n);
	vector<double> weight(entity_neighbors[make_pair(e, rel)].size());
	double sum = 0;
    	for (int i = 0; i<entity_neighbors[make_pair(e, rel)].size(); i++)
	{
		sum += exp(get_sum(e, entity_neighbors[make_pair(e, rel)][i], rel));
		weight.push_back(exp(get_sum(e, entity_neighbors[make_pair(e, rel)][i], rel)));
	}
	for (int ii=0; ii<n; ii++)
	{
		attention_vec[ii] = 0.0;
		for (int i=0; i<entity_neighbors[make_pair(e, rel)].size(); i++)
			attention_vec[ii] += weight[i]/sum*entity_vec[entity_neighbors[make_pair(e, rel)][i]][ii];
	}
	return attention_vec;
    }
    double get_sum(int e1, int e2, int rel)
    {
	double sum=0;
	if (L1_flag)
		for (int ii=0; ii<n; ii++)
			sum+=fabs(entity_vec[e2][ii]-entity_vec[e1][ii]-relation_vec[rel][ii]);
	else
		for (int ii=0; ii<n; ii++)
			sum+=sqr(entity_vec[e2][ii]-entity_vec[e1][ii]-relation_vec[rel][ii]);
	return sum;
    }
	
    double calc_sum(int e1,int e2,int rel)
    {
	vector<double> neighbors_e1 = get_neighbors_attention_vec(e1, rel);
	vector<double> neighbors_e2 = get_neighbors_attention_vec(e2, rel);
        double sum=0;
        if (L1_flag)
        	for (int ii=0; ii<n; ii++)
            	sum+=fabs(entity_vec[e2][ii]-entity_vec[e1][ii]-relation_vec[rel][ii]) + fabs(neighbors_e2[ii]-neighbors_e1[ii]-relation_vec[rel][ii] + con_alpha*(neighbors_e2[ii]-entity_vec[e2][ii] + neighbors_e1[ii]-entity_vec[e1][ii]));
        else
        	for (int ii=0; ii<n; ii++)
            	sum+=sqr(entity_vec[e2][ii]-entity_vec[e1][ii]-relation_vec[rel][ii]) + sqr(neighbors_e2[ii]-neighbors_e1[ii]-relation_vec[rel][ii] + con_alpha*(neighbors_e2[ii]-entity_vec[e2][ii] + neighbors_e1[ii]-entity_vec[e1][ii]));
        return sum;
    }
    void gradient(int e1_a,int e2_a,int rel_a,int e1_b,int e2_b,int rel_b)
    {
	vector<double> neighbors_e1_a = get_neighbors_attention_vec(e1_a, rel_a);
	vector<double> neighbors_e2_a = get_neighbors_attention_vec(e2_a, rel_a);
        vector<double> neighbors_e1_b = get_neighbors_attention_vec(e1_b, rel_b);
        vector<double> neighbors_e2_b = get_neighbors_attention_vec(e2_b, rel_b);
        for (int ii=0; ii<n; ii++)
        {

            double x = 2*(entity_vec[e2_a][ii]-entity_vec[e1_a][ii]-relation_vec[rel_a][ii]+neighbors_e2_a[ii]-neighbors_e1_a[ii]-relation_vec[rel_a][ii]+con_alpha*(neighbors_e2_a[ii]-entity_vec[e2_a][ii]+neighbors_e1_a[ii]-entity_vec[e1_a][ii]));
            if (L1_flag)
            	if (x>0)
            		x=1;
            	else
            		x=-1;
            relation_tmp[rel_a][ii]-=-1*rate*x;
            entity_tmp[e1_a][ii]-=-1*rate*x;
            entity_tmp[e2_a][ii]+=-1*rate*x;
            x = 2*(entity_vec[e2_b][ii]-entity_vec[e1_b][ii]-relation_vec[rel_b][ii]+neighbors_e2_b[ii]-neighbors_e1_b[ii]-relation_vec[rel_b][ii]+con_alpha*(neighbors_e2_b[ii]-entity_vec[e2_b][ii]+neighbors_e1_b[ii]-entity_vec[e2_a][ii]));
            if (L1_flag)
            	if (x>0)
            		x=1;
            	else
            		x=-1;
            relation_tmp[rel_b][ii]-=rate*x;
            entity_tmp[e1_b][ii]-=rate*x;
            entity_tmp[e2_b][ii]+=rate*x;
        }
    }
    void train_kb(int e1_a,int e2_a,int rel_a,int e1_b,int e2_b,int rel_b)
    {
        double sum1 = calc_sum(e1_a,e2_a,rel_a);
        double sum2 = calc_sum(e1_b,e2_b,rel_b);
        if (sum1+margin>sum2)
        {
        	res+=margin+sum1-sum2;
        	gradient( e1_a, e2_a, rel_a, e1_b, e2_b, rel_b);
        }
    }
};

Train train;
void prepare()
{
    FILE* f1 = fopen((data_path+"entity2id.txt").c_str(),"r");
    FILE* f2 = fopen((data_path+"relation2id.txt").c_str(),"r");
    int x;
    while (fscanf(f1,"%s%d",buf,&x)==2)
    {
	string st=buf;
	entity2id[st]=x;
	id2entity[x]=st;
	entity_num++;
    }
    while (fscanf(f2,"%s%d",buf,&x)==2)
    {
	string st=buf;
	relation2id[st]=x;
	id2relation[x]=st;
	relation_num++;
    }
    FILE* f_kb = fopen((data_path+"train.txt").c_str(),"r");
    int cnt = 0;
	while (fscanf(f_kb,"%s",buf)==1)
    {
        string s1=buf;
        fscanf(f_kb,"%s",buf);
        string s2=buf;
        fscanf(f_kb,"%s",buf);
        string s3=buf;
        if (entity2id.count(s1)==0)
        {
            cout<<"miss entity:"<<s1<<endl;
        }
        if (entity2id.count(s2)==0)
        {
            cout<<"miss entity:"<<s2<<endl;
        }
        if (relation2id.count(s3)==0)
        {
            relation2id[s3] = relation_num;
            relation_num++;
	    cnt++;
        }
        left_entity[relation2id[s3]][entity2id[s1]]++;
        right_entity[relation2id[s3]][entity2id[s2]]++;
        train.add(entity2id[s1],entity2id[s2],relation2id[s3]);
    }
    for (int i=0; i<relation_num; i++)
    {
    	double sum1=0,sum2=0;
    	for (map<int,int>::iterator it = left_entity[i].begin(); it!=left_entity[i].end(); it++)
    	{
    		sum1++;
    		sum2+=it->second;
    	}
    	left_num[i]=sum2/sum1;
    }
    for (int i=0; i<relation_num; i++)
    {
    	double sum1=0,sum2=0;
    	for (map<int,int>::iterator it = right_entity[i].begin(); it!=right_entity[i].end(); it++)
    	{
    		sum1++;
    		sum2+=it->second;
    	}
    	right_num[i]=sum2/sum1;
	//cout << right_num[i] << " ";
    }
//    for (map<int, int>::iterator it = right_entity[0].begin(); it!=right_entity[0].end(); it++)
//	cout << it->first << ":" << it->second << ",";
    cout<<"relation_num="<<relation_num<<endl;
    cout<<"entity_num="<<entity_num<<endl;
    fclose(f_kb);
}

int ArgPos(char *str, int argc, char **argv) {
  int a;
  for (a = 1; a < argc; a++) if (!strcmp(str, argv[a])) {
    if (a == argc - 1) {
      printf("Argument missing for %s\n", str);
      exit(1);
    }
    return a;
  }
  return -1;
}

int main(int argc,char**argv)
{
    srand((unsigned) time(NULL));
    int method = 1;
    int n = 100;
    double rate = 0.001;
    double margin = 1;
    double con_alpha = 0.01;
    int i;
    if ((i = ArgPos((char *)"-size", argc, argv)) > 0) n = atoi(argv[i + 1]);
    if ((i = ArgPos((char *)"-margin", argc, argv)) > 0) margin = atoi(argv[i + 1]);
    if ((i = ArgPos((char *)"-method", argc, argv)) > 0) method = atoi(argv[i + 1]);
    if ((i = ArgPos((char *)"-con_alpha", argc, argv)) > 0) con_alpha = atoi(argv[i + 1]);
    cout<<"size = "<<n<<endl;
    cout<<"learing rate = "<<rate<<endl;
    cout<<"margin = "<<margin<<endl;
    if (method)
        version = "bern";
    else
        version = "unif";
    cout<<"method = "<<version<<endl;
    cout<<"con_alpha = "<<con_alpha<<endl;
    prepare();
    train.run(n,rate,margin,method, con_alpha);
}
